File: bffs/mobile-bff/app/routers/rental_routes.py
from fastapi import APIRouter, HTTPException, Query
import httpx

from app.clients import rental_client
from app.schemas.mobile_schemas import UpdateRentalStatusRequest

router = APIRouter(prefix="/api/mobile/rentals", tags=["Mobile Rentals"])


@router.get("/my/{user_id}")
async def my_rentals(user_id: str):
    try:
        as_borrower = await rental_client.get_rentals_by_borrower(user_id)
        as_owner = await rental_client.get_rentals_by_owner(user_id)

        return {
            "asBorrower": [
                {
                    "id": rental.get("id"),
                    "itemId": rental.get("itemId"),
                    "status": rental.get("status"),
                    "startDate": rental.get("startDate"),
                    "endDate": rental.get("endDate"),
                }
                for rental in as_borrower
            ],
            "asOwner": [
                {
                    "id": rental.get("id"),
                    "itemId": rental.get("itemId"),
                    "borrowerId": rental.get("borrowerId"),
                    "status": rental.get("status"),
                    "startDate": rental.get("startDate"),
                    "endDate": rental.get("endDate"),
                }
                for rental in as_owner
            ],
        }
    except httpx.HTTPStatusError as ex:
        raise HTTPException(
            status_code=ex.response.status_code,
            detail=ex.response.json()
        )


@router.patch("/{rental_id}/status")
async def update_status(rental_id: str, request: UpdateRentalStatusRequest):
    try:
        return await rental_client.update_rental_status(
            rental_id,
            request.model_dump(exclude_none=True)
        )
    except httpx.HTTPStatusError as ex:
        raise HTTPException(
            status_code=ex.response.status_code,
            detail=ex.response.json()
        )


@router.delete("/{rental_id}")
async def cancel_rental(rental_id: str, requesterId: str = Query(...)):
    try:
        return await rental_client.cancel_rental(rental_id, requesterId)
    except httpx.HTTPStatusError as ex:
        raise HTTPException(
            status_code=ex.response.status_code,
            detail=ex.response.json()
        )

File: bffs/web-bff/src/routes/rentalRoutes.js
const express = require("express");
const rentalClient = require("../clients/rentalClient");

const router = express.Router();

router.get("/", async (req, res, next) => {
  try {
    const status = typeof req.query.status === "string" ? req.query.status : undefined;
    const rentals = await rentalClient.getRentals(status);
    return res.json(rentals);
  } catch (error) {
    return next(error);
  }
});

router.get("/borrower/:borrowerId", async (req, res, next) => {
  try {
    const rentals = await rentalClient.getRentalsByBorrower(req.params.borrowerId);
    return res.json(rentals);
  } catch (error) {
    return next(error);
  }
});

router.get("/owner/:ownerId", async (req, res, next) => {
  try {
    const rentals = await rentalClient.getRentalsByOwner(req.params.ownerId);
    return res.json(rentals);
  } catch (error) {
    return next(error);
  }
});

router.get("/item/:itemId", async (req, res, next) => {
  try {
    const rentals = await rentalClient.getRentalsByItem(req.params.itemId);
    return res.json(rentals);
  } catch (error) {
    return next(error);
  }
});

router.get("/:id", async (req, res, next) => {
  try {
    const rental = await rentalClient.getRental(req.params.id);
    return res.json(rental);
  } catch (error) {
    return next(error);
  }
});

router.post("/", async (req, res, next) => {
  try {
    const rental = await rentalClient.createRental(req.body);
    return res.status(201).json(rental);
  } catch (error) {
    return next(error);
  }
});

router.patch("/:id/status", async (req, res, next) => {
  try {
    const rental = await rentalClient.updateRentalStatus(req.params.id, req.body);
    return res.json(rental);
  } catch (error) {
    return next(error);
  }
});

router.delete("/:id", async (req, res, next) => {
  try {
    const requesterId =
      typeof req.query.requesterId === "string" ? req.query.requesterId : undefined;

    if (!requesterId) {
      return res.status(400).json({
        error: "requesterId query parameter is required"
      });
    }

    const result = await rentalClient.cancelRental(req.params.id, requesterId);
    return res.json(result);
  } catch (error) {
    return next(error);
  }
});

module.exports = router;